apps/mark-scan/backend/src/pat-input/patinputd.c

```diff
@@ -0,0 +1,170 @@
+  ioctl(uinput_fd, UI_SET_EVBIT, EV_KEY);
+  ioctl(uinput_fd, UI_SET_KEYBIT, KEY_1);
+  ioctl(uinput_fd, UI_SET_KEYBIT, KEY_2);
+
+  memset(&usetup, 0, sizeof(usetup));
+  usetup.id.bustype = BUS_USB;
+  usetup.id.vendor = 0x1234;  /* sample vendor */
+  usetup.id.product = 0x5678; /* sample product */
+  strcpy(usetup.name, "PAT Input daemon virtual device");
+
+  ioctl(uinput_fd, UI_DEV_SETUP, &usetup);
+  ioctl(uinput_fd, UI_DEV_CREATE);
+
+  /*
+   * On UI_DEV_CREATE the kernel will create the device node for this
+   * device. We are inserting a pause here so that userspace has time
+   * to detect, initialize the new device, and can start listening to
+   * the event, otherwise it will not notice the event we are about
+   * to send. This pause is only needed in our example code!
+   */
+  sleep(1);
+
+  // TODO handle already-exported pins
+  export_pin(pat_is_connected_gpio_number);
+  export_pin(pat_a_signal_gpio_number);
+  export_pin(pat_b_signal_gpio_number);
+
+  set_pin_direction_in(pat_is_connected_gpio_number);
+  set_pin_direction_in(pat_a_signal_gpio_number);
+  set_pin_direction_in(pat_b_signal_gpio_number);
+
+  int is_connected_value_fd = get_pin_value_fd(pat_is_connected_gpio_number);
+  bool is_connected = get_bool_pin_value(is_connected_value_fd);
+
+  printf("Initial PAT device is connected: %s\n", is_connected ? "true" : "false");
+
+  int a_signal_fd = get_pin_value_fd(pat_a_signal_gpio_number);
+  bool a_signal = get_bool_pin_value(a_signal_fd);
+  printf("Initial A value: %s\n", a_signal ? "true" : "false");
+
+  int b_signal_fd = get_pin_value_fd(pat_b_signal_gpio_number);
+  bool b_signal = get_bool_pin_value(b_signal_fd);
+  printf("Initial B value: %s\n", b_signal ? "true" : "false");
+
+  signal(SIGINT, interrupt);
+
+  while (!should_exit_cleanly)
+  {
+    // Need to get file descriptor each time we read or the value will be stale
+    a_signal_fd = get_pin_value_fd(pat_a_signal_gpio_number);
+    bool new_a_signal = get_bool_pin_value(a_signal_fd);
+
+    b_signal_fd = get_pin_value_fd(pat_b_signal_gpio_number);
+    bool new_b_signal = get_bool_pin_value(b_signal_fd);
+
+    // Only emit keyboard event when signal changes
+    if (new_a_signal && !a_signal)
+    {
+      printf("'A' signal triggered\n");
+      /* Key press, report the event, send key release, and report again */
+      emit(uinput_fd, EV_KEY, KEY_1, 1);
+      emit(uinput_fd, EV_SYN, SYN_REPORT, 0);
+      emit(uinput_fd, EV_KEY, KEY_1, 0);
+      emit(uinput_fd, EV_SYN, SYN_REPORT, 0);
+    }
+
+    if (new_b_signal && !b_signal)
+    {
+      printf("'B' signal triggered\n");
+      /* Key press, report the event, send key release, and report again */
+      emit(uinput_fd, EV_KEY, KEY_2, 1);
+      emit(uinput_fd, EV_SYN, SYN_REPORT, 0);
+      emit(uinput_fd, EV_KEY, KEY_2, 0);
+      emit(uinput_fd, EV_SYN, SYN_REPORT, 0);
+    }
+
+    a_signal = new_a_signal;
+    b_signal = new_b_signal;
+    int close_result = close(a_signal_fd);
```

Comment:
any reason you're reusing `close_result` rather than having `a_close_result` and `b_close_result` as with other vars?

Reply:
I had allocated these for debugging but I think you're right they don't need to be allocated at all now.

Reply:
[75a0ca5](https://github.com/votingworks/vxsuite/pull/4210/commits/75a0ca5497dc5c298a35276545e3036cfe1a1b90)